<h1>Data Wrangling</h1>

<h2>What is the purpose of Data Wrangling?</h2>

Data Wrangling is the process of converting data from the initial format to a format that may be better for analysis.

<h3>What is the fuel consumption (L/100k) rate for the diesel car?</h3>

<h3>Import data</h3>
<p>
You can find the "Automobile Data Set" from the following: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos"> Link</a>. 
We will be using this data set throughout this course.
</p>

<h4>Import pandas</h4> 

In [1]:
import pandas as pd
#import matplotlib.pylab as plt

<h2>Reading the data set from the URL and adding the related headers.</h2>

In [2]:
# Read the online file by the URL provides above, and assign it to variable "df"
path = "./auto.csv"
df = pd.read_csv(path, header=None)

In [3]:
df.head()

0    1            2    3    4     5            6    7      8     9   ...    \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...     
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...     
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...     
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...     
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...     

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

 Python list <b>headers</b> containing name of headers 

In [4]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

df.columns = headers
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

Use the Pandas method <b>read_csv()</b> to load the data from the web address. Set the parameter  "names" equal to the Python list "headers".

<h2 id="identify_handle_missing_values">Identify and handle missing values</h2>
<ol>
    <li>Remove Empty Value</li>
    <li>Remove NaN</li>
</ol>


In [5]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...    engine-size  \
0  convertible          rwd           front        88.6  ...            130   
1  convertible          rwd           front        88.6  ...            130   
2    hatchback          rwd           front        94.5  ...            152   
3        sedan          fwd           front        99.8  ...            109   
4        sedan          4wd           front        99.4  ...            136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [6]:
df.isnull()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0        False              False  False      False       False         False   
1        False              False  False      False       False         False   
2        False              False  False      False       False         False   
3        False              False  False      False       False         False   
4        False              False  False      False       False         False   
5        False              False  False      False       False         False   
6        False              False  False      False       False         False   
7        False              False  False      False       False         False   
8        False              False  False      False       False         False   
9        False              False  False      False       False         False   
10       False              False  False      False       False         False   
11       False              False  False      False       False         False   
12       False              False  False      False       False         False   
13       False              False  False      False       False         False   
14       False              False  False      False       False         False   
15       False              False  False      False       False         False   
16       False              False  False      False       False         False   
17       False              False  False      False       False         False   
18       False              False  False      False       False         False   
19       False              False  False      False       False         False   
20       False              False  False      False       False         False   
21       False              False  False      False       False         False   
22       False              False  False      False       False         False   
23       False              False  False      False       False         False   
24       False              False  False      False       False         False   
25       False              False  False      False       False         False   
26       False              False  False      False       False         False   
27       False              False  False      False       False         False   
28       False              False  False      False       False         False   
29       False              False  False      False       False         False   
..         ...                ...    ...        ...         ...           ...   
175      False              False  False      False       False         False   
176      False              False  False      False       False         False   
177      False              False  False      False       False         False   
178      False              False  False      False       False         False   
179      False              False  False      False       False         False   
180      False              False  False      False       False         False   
181      False              False  False      False       False         False   
182      False              False  False      False       False         False   
183      False              False  False      False       False         False   
184      False              False  False      False       False         False   
185      False              False  False      False       False         False   
186      False              False  False      False       False         False   
187      False              False  False      False       False         False   
188      False              False  False      False       False         False   
189      False              False  False      False       False         False   
190      False              False  False      False       False         False   
191      False              False  False      False       False         False   
192      False              False  F

As we can see, several question marks appeared in the dataframe; those are missing values which may hinder our further analysis. 
<div>So, how do we identify all those missing values and deal with them?</div> 


<b>How to work with missing data?</b>

Steps for working with missing data:
<ol>
    <li>identify missing data</li>
    <li>deal with missing data</li>
    <li>correct data format</li>
</ol>

<h3 id="identify_missing_values">Identify missing values</h3>
<h4>Convert "?" to NaN</h4>
In the car dataset, missing data comes with the question mark "?".
We replace "?" with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience. Here we use the function: 
 <pre>.replace(A, B, inplace = True) </pre>
to replace A by B

In [7]:
import numpy as np

# replace "?" to NaN
df.replace("?", np.nan, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...    engine-size  \
0  convertible          rwd           front        88.6  ...            130   
1  convertible          rwd           front        88.6  ...            130   
2    hatchback          rwd           front        94.5  ...            152   
3        sedan          fwd           front        99.8  ...            109   
4        sedan          4wd           front        99.4  ...            136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

dentify_missing_values

<h4>Evaluating for Missing Data</h4>

The missing values are converted to Python's default. We use Python's built-in functions to identify these missing values. There are two methods to detect missing data:
<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [8]:
missing_data = df.isnull()
missing_data.head(5)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...    engine-size  \
0       False         False            False       False  ...          False   
1       False         False            False       False  ...          False   
2       False         False            False       False  ...          False   
3       False         False            False       False  ...          False   
4       False         False            False       False  ...          False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

"True" stands for missing value, while "False" stands for not missing value.

<h4>Count missing values in each column</h4>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False"  means the value is present in the dataset.  In the body of the for loop the method  ".value_counts()"  counts the number of "True" values. 
</p>

In [9]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")    

symboling
False    205
Name: symboling, dtype: int64

normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

make
False    205
Name: make, dtype: int64

fuel-type
False    205
Name: fuel-type, dtype: int64

aspiration
False    205
Name: aspiration, dtype: int64

num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

body-style
False    205
Name: body-style, dtype: int64

drive-wheels
False    205
Name: drive-wheels, dtype: int64

engine-location
False    205
Name: engine-location, dtype: int64

wheel-base
False    205
Name: wheel-base, dtype: int64

length
False    205
Name: length, dtype: int64

width
False    205
Name: width, dtype: int64

height
False    205
Name: height, dtype: int64

curb-weight
False    205
Name: curb-weight, dtype: int64

engine-type
False    205
Name: engine-type, dtype: int64

num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64

engine-size
False    205
Name: engine-size, dtype: int64

fuel-system
Fa

**Or simply by using the sum fuction**

In [10]:
missing_data.sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

Based on the summary above, each column has 205 rows of data, seven columns containing missing data:
<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
        c. replace it based on other functions
    </li>
</ol>

Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely.
We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

<b>Replace by mean:</b>
<ul>
    <li>"normalized-losses": 41 missing data, replace them with mean</li>
    <li>"stroke": 4 missing data, replace them with mean</li>
    <li>"bore": 4 missing data, replace them with mean</li>
    <li>"horsepower": 2 missing data, replace them with mean</li>
    <li>"peak-rpm": 2 missing data, replace them with mean</li>
</ul>

<b>Replace by frequency:</b>
<ul>
    <li>"num-of-doors": 2 missing data, replace them with "four". 
        <ul>
            <li>Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur</li>
        </ul>
    </li>
</ul>

<b>Drop the whole row:</b>
<ul>
    <li>"price": 4 missing data, simply delete the whole row
        <ul>
            <li>Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us</li>
        </ul>
    </li>
</ul>

<h4>Calculate the average of the column </h4>

In [11]:
avg_norm_loss = df["normalized-losses"].astype("float").mean(axis=0)
print("Average of normalized-losses:", avg_norm_loss)

Average of normalized-losses: 122.0


<h4>Replace "NaN" by mean value in "normalized-losses" column</h4>

In [12]:
df["normalized-losses"].replace(np.nan, avg_norm_loss, inplace=True)
df["normalized-losses"].isnull().sum()

0

Double-click <b>here</b> for the solution.

<!-- The answer is below:

# calculate the mean vaule for "stroke" column
avg_stroke = df["stroke"].astype("float").mean(axis = 0)
print("Average of stroke:", avg_stroke)

# replace NaN by mean value in "stroke" column
df["stroke"].replace(np.nan, avg_stroke, inplace = True)

-->


To see which values are present in a particular column, we can use the ".value_counts()" method:

In [13]:
df['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate for us the most common type automatically:

In [14]:
df['num-of-doors'].value_counts().idxmax()

'four'

The replacement procedure is very similar to what we have seen previously

In [15]:
df["num-of-doors"].replace(np.nan, "four", inplace=True)

Finally, let's drop all rows that do not have price data:

In [16]:
df.dropna(subset=["price"], axis=0, inplace=True)
df.reset_index(drop=True, inplace=True)

In [17]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               122  alfa-romero       gas        std          two   
1          3               122  alfa-romero       gas        std          two   
2          1               122  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...    engine-size  \
0  convertible          rwd           front        88.6  ...            130   
1  convertible          rwd           front        88.6  ...            130   
2    hatchback          rwd           front        94.5  ...            152   
3        sedan          fwd           front        99.8  ...            109   
4        sedan          4wd           front        99.4  ...            136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

<b>Good!</b> Now, we obtain the dataset with no missing values.

<h3 id="correct_data_format">Correct data format</h3>
<b>We are almost there!</b>
<p>The last step in data cleaning is checking and making sure that all data is in the correct format (int, float, text or other).</p>

In Pandas, we use 
<p><b>.dtype()</b> to check the data type</p>
<p><b>.astype()</b> to change the data type</p>

<h4>Lets list the data types for each column</h4>

In [18]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

<p>As we can see above, some columns are not of the correct data type. Numerical variables should have type 'float' or 'int', and variables with strings such as categories should have type 'object'. For example, 'bore' and 'stroke' variables are numerical values that describe the engines, so we should expect them to be of the type 'float' or 'int'; however, they are shown as type 'object'. We have to convert data types into a proper format for each column using the "astype()" method.</p> 

<h4>Convert data types to proper format</h4>

In [19]:
df["bore"] = df["bore"].astype("float")
df['bore'].dtype

dtype('float64')

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  #1: </h1>

<b>According to the example above, replace NaN in "stroke" ,"bore" ,"horsepower"  ,"peak-rpm" column by mean.</b>
</div>

In [14]:

import pandas as pd
df = pd.read_csv("D:\data.csv")
#df.mean(axis = 1, skipna = True) 
df['stroke'].mean()
df.describe()



Unnamed: 0   symboling  normalized-losses  wheel-base      length  \
count  205.000000  205.000000         164.000000  205.000000  205.000000   
mean   102.000000    0.834146         122.000000   98.756585  174.049268   
std     59.322565    1.245307          35.442168    6.021776   12.337289   
min      0.000000   -2.000000          65.000000   86.600000  141.100000   
25%     51.000000    0.000000          94.000000   94.500000  166.300000   
50%    102.000000    1.000000         115.000000   97.000000  173.200000   
75%    153.000000    2.000000         150.000000  102.400000  183.100000   
max    204.000000    3.000000         256.000000  120.900000  208.100000   

            width      height  curb-weight  engine-size        bore  \
count  205.000000  205.000000   205.000000   205.000000  201.000000   
mean    65.907805   53.724878  2555.565854   126.907317    3.329751   
std      2.145204    2.443522   520.680204    41.642693    0.273539   
min     60.300000   47.800000  1488.000000    61.000000    2.540000   
25%     64.100000   52.000000  2145.000000    97.000000    3.150000   
50%     65.500000   54.100000  2414.000000   120.000000    3.310000   
75%     66.900000   55.500000  2935.000000   141.000000    3.590000   
max     72.300000   59.800000  4066.000000   326.000000    3.940000   

           stroke  compression-ratio  horsepower     peak-rpm    city-mpg  \
count  201.000000         205.000000  203.000000   203.000000  205.000000   
mean     3.255423          10.142537  104.256158  5125.369458   25.219512   
std      0.316717           3.972040   39.714369   479.334560    6.542142   
min      2.070000           7.000000   48.000000  4150.000000   13.000000   
25%      3.110000           8.600000   70.000000  4800.000000   19.000000   
50%      3.290000           9.000000   95.000000  5200.000000   24.000000   
75%      3.410000           9.400000  116.000000  5500.000000   30.000000   
max      4.170000          23.000000  288.000000  6600.000000   49.000000   

       highway-mpg         price  
count   205.000000    201.000000  
mean     30.751220  13207.129353  
std       6.886443   7947.066342  
min      16.000000   5118.000000  
25%      25.000000   7775.000000  
50%      30.000000  10295.000000  
75%      34.000000  16500.000000  
max      54.000000  45400.000000

In [16]:

import numpy as np
avg_norm_loss = df["stroke"].astype("float").mean(axis=0)
print("Average of strokes:", avg_norm_loss)
df["stroke"].replace(np.nan, avg_norm_loss, inplace=True)
df["stroke"].isnull().sum()

Average of normalized-losses: 3.2554228855721337


0

In [17]:
import numpy as np
avg_norm_loss = df["bore"].astype("float").mean(axis=0)
print("Average of bore:", avg_norm_loss)
df["bore"].replace(np.nan, avg_norm_loss, inplace=True)
df["bore"].isnull().sum()

Average of bore: 3.3297512437810957


0

In [19]:
import numpy as np
avg_norm_loss = df["horsepower"].astype("float").mean(axis=0)
print("Average of horsepower:", avg_norm_loss)
df["horsepower"].replace(np.nan, avg_norm_loss, inplace=True)
df["horsepower"].isnull().sum()

Average of horsepower: 104.25615763546799


0

In [20]:
import numpy as np
avg_norm_loss = df["peak-rpm"].astype("float").mean(axis=0)
print("Average of peak-rpm:", avg_norm_loss)
df["peak-rpm"].replace(np.nan, avg_norm_loss, inplace=True)
df["peak-rpm"].isnull().sum()

Average of peak-rpm: 5125.369458128079


0

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question  #2: </h1>

<b>According to the example above, correct the datatypes of "normalized-losses" "price" & "peak-rpm".</b>
</div>

In [9]:
df["normalized-losses"] = df["normalized-losses"].astype("object")
df['normalized-losses'].dtype

dtype('O')

In [10]:
df["price"] = df["price"].astype("object")
df['price'].dtype

dtype('O')

In [11]:
df["peak-rpm"] = df["peak-rpm"].astype("object")
df['peak-rpm'].dtype

dtype('O')